docs/AddressSpace.md

```diff
@@ -0,0 +1,153 @@
+
+   1. A `RemoteAllocator*` indicating the `LocalAlloc` that owns the region of memory.
+
+   2. A "full sizeclass" value (itself a tagged sum type between large and small sizeclasses).
+
+2. The `meta` pointer is a bit-stuffed pair of
+
+   1. A pointer to a larger metadata structure with type dependent on the role of this chunk
+
+   2. A bit (`META_BOUNDARY_BIT`) that serves to limit chunk coalescing on platforms where that may not be possible, such as CHERI.
+
+See `src/backend/metatypes.h` and `src/mem/metaslab.h`.
+
+For chunks owned by a *backend* (`REMOTE_BACKEND_MARKER` asserted), there are again multiple possibilities.
+
+For chunks owned by a *small buddy allocator*, the remainder of the `MetaEntry` is zero.
+That is, it appears to have small sizeclass 0 and an implausible `RemoteAllocator*`.
+
+For chunks owned by a *large buddy allocator*, the `MetaEntry` is instead a node in a red-black tree of all such chunks.
+Its contents can be decoded as follows:
+
+1. The `meta` field's `META_BOUNDARY_BIT` is preserved, with the same meaning as in the frontend case, above.
+
+2. `meta` (resp. `remote_and_sizeclass`) includes a pointer to the left (resp. right) *chunk* of address space.
+   (The corresponding child *node* in this tree is found by taking the *address* of this chunk and looking up the `MetaEntry` in the Pagemap.)
+
+3. The `meta` field's `LargeBuddyRep::RED_BIT` is used to carry the red/black color of this node.
+
+See `src/backend/largebuddyrange.h`.
+
+### Encoding a MetaEntry
+
+We can also consider the process for generating a MetaEntry for a chunk of the address space given its state.
+The following cases apply:
+
+1. The address is not associated with `snmalloc`:
+   Here, the `MetaEntry`, if it is mapped, is all zeros and so it...
+   * has `REMOTE_BACKEND_MARKER` clear in `remote_and_sizeclass`.
+   * appears to be owned by a frontend RemoteAllocator at address 0 (probably, but not certainly, `nullptr`).
+   * has "small" sizeclass 0, which has size 0.
+   * has no associated metadata structure.
+
+2. The address is part of a free chunk in a backend's Large Buddy Allocator:
+   The `MetaEntry`...
+   * has `REMOTE_BACKEND_MARKER` asserted in `remote_and_sizeclass`.
+   * has "small" sizeclass 0, which has size 0.
+   * the remainder of its `MetaEntry` structure will be a Large Buddy Allocator rbtree node.
+   * has no associated metadata structure.
+
+3. The address is part of a free chunk inside a backend's Small Buddy Allocator:
+   Here, the `MetaEntry` is zero aside from the asserted `REMOTE_BACKEND_MARKER` bit, and so it...
+   * has "small" sizeclass 0, which has size 0.
+   * has no associated metadata structure.
+
+4. The address is part of a live large allocation (spanning one or more 16KiB chunks):
+   Here, the `MetaEntry`...
+   * has `REMOTE_BACKEND_MARKER` clear in `remote_and_sizeclass`.
+   * has a *large* sizeclass value.
+   * has an associated `RemoteAllocator*` and metadata structure.
```

Comment:
What is the type of the metadata structure?

Reply:
I think it's currently a `ChunkRecord`, but that type isn't long for the world, and I think the `next` pointer therein is unused, making it equivalent to a naked `MetaCommon`.  If that's right, @mjp41, should I add a commit that does away with the use of `ChunkRecord` for this case?

Reply:
Correct. It is a Metaslab, that always triggers a slow path on deallocation, so we can detect large off the fast path.